scripts/rewards_complex.py

```diff
@@ -0,0 +1,187 @@
+        self.root.distribute(reward)
+
+    def withdraw_reward(self, root_key, leaf_key):
+        reward = self.root.withdraw_reward(root_key)
+        self.leaf[root_key].distribute(reward)
+        return self.leaf[root_key].withdraw_reward(leaf_key)
+
+
+class VaultRegistry:
+    def __init__(self):
+        self.exchange_rate = {}
+        self.secure_threshold = {}
+
+        self.rewards = RewardsRouter()
+
+        self.collateral = {}
+        self.total_collateral_div_threshold = {}
+
+    # relative to btc
+    def set_exchange_rate(self, currency, value):
+        if currency not in self.exchange_rate:
+            self.exchange_rate[currency] = value
+            return
+
+        previous_collateral_capacity = self.get_collateral_capacity(currency)
+        self.exchange_rate[currency] = value
+        collateral_capacity = self.get_collateral_capacity(currency)
+        capacity_delta = collateral_capacity - previous_collateral_capacity
+        self.rewards.update_root_stake(currency, capacity_delta)
+
+    def set_global_secure_threshold(self, currency, value):
+        # this is tricky to update without summing
+        self.secure_threshold[currency] = value
+
+    def get_secure_threshold(self, vault):
+        return max(self.secure_threshold[vault.currency], vault.secure_threshold)
+
+    def update_collateral_and_threshold(self, vault, amount, secure_threshold):
+        currency = vault.currency
+
+        if secure_threshold is None:
+            secure_threshold = self.get_secure_threshold(vault)
+
+        previous_collateral_capacity = self.get_collateral_capacity(currency)
+        previous_secure_threshold = self.get_secure_threshold(vault)
+        previous_rate = self.collateral[vault] / previous_secure_threshold
+
+        self.collateral[vault] += amount
+        next_rate = self.collateral[vault] / secure_threshold - previous_rate
+        self.total_collateral_div_threshold[currency] += next_rate
+
+        collateral_capacity = self.get_collateral_capacity(currency)
+        capacity_delta = collateral_capacity - previous_collateral_capacity
+
+        self.rewards.update_stake(
+            currency, capacity_delta, vault.address, amount)
+
+    def set_custom_secure_threshold(self, vault, value):
+        secure_threshold = max(self.secure_threshold[vault.currency], value)
+        self.update_collateral_and_threshold(vault, 0, secure_threshold)
+        vault.set_secure_threshold(value)
+
+    def deposit_collateral(self, vault, amount):
+        currency = vault.currency
+        if vault not in self.collateral:
+            self.collateral[vault] = 0
+        if currency not in self.total_collateral_div_threshold:
+            self.total_collateral_div_threshold[currency] = 0
+
+        self.update_collateral_and_threshold(vault, amount, None)
+
+    def get_collateral_capacity(self, currency):
+        return self.total_collateral_div_threshold[currency] \
+            / self.exchange_rate[currency]
```

Comment:
If the exchange rate is in base quotation (1 DOT = x USD) and your accounting currency is the base currency (e.g DOT) than you have to multiply by the exchange rate.

Reply:
In the parachain the unit is plancks per satoshi

Reply:
Okay, all good then.